src/drt/src/dr/FlexDR.cpp

```diff
@@ -1230,6 +1235,167 @@ void FlexDR::reportGuideCoverage()
+  std::sort(lonely_vias_regions.begin(),
+            lonely_vias_regions.end(),
+            [](const Rect& a, const Rect& b) { return a.xMin() < b.xMin(); });
+  std::vector<Rect> merged_regions;
+  for (const auto& region : lonely_vias_regions) {
+    bool found = false;
+    for (auto& merged_region : merged_regions) {
+      if (region.intersects(merged_region)) {
+        merged_region.merge(region);
+        found = true;
+        break;
+      }
+    }
+    if (!found) {
+      merged_regions.emplace_back(region);
+    }
+  }
+  // create drWorkers for the final regions
+  omp_set_num_threads(MAX_THREADS);
+#pragma omp parallel for schedule(dynamic)
+  for (size_t i = 0; i < merged_regions.size(); i++) {
+    auto route_box = merged_regions.at(i);
+    auto worker = std::make_unique<FlexDRWorker>(&via_data_, design_, logger_);
+    Rect ext_box;
+    Rect drc_box;
+    auto minGcellIdx = getDesign()->getTopBlock()->getGCellIdx(
+        {route_box.xMin(), route_box.yMin()});
+    route_box.bloat(MTSAFEDIST, ext_box);
+    route_box.bloat(DRCSAFEDIST, drc_box);
+    worker->setRouteBox(route_box);
+    worker->setExtBox(ext_box);
+    worker->setDrcBox(drc_box);
+    worker->setDRIter(64);
+    worker->setDebugSettings(router_->getDebugSettings());
+    worker->setRipupMode(RipUpMode::VIASWAP);
+    worker->setGraphics(graphics_.get());
+    worker->main(getDesign());
+#pragma omp critical
+    {
+      worker->end(getDesign());
+    }
+  }
+}
+
+std::vector<frVia*> FlexDR::getLonelyVias(frLayer* layer,
+                                          int max_spc,
+                                          int cut_class)
+{
+  std::vector<frVia*> lonely_vias;
+  if (layer->getSecondaryViaDefs().empty()) {
+    return lonely_vias;
+  }
+  auto vias = getRegionQuery()->getVias(layer->getLayerNum());
+  std::vector<std::pair<int, int>> via_positions;
+  via_positions.reserve(vias.size());
+  for (auto [obj, box] : vias) {
+    via_positions.emplace_back(box.xCenter(), box.yCenter());
+  }
+  KDTree tree(via_positions);
+  std::vector<bool> visited(via_positions.size(), false);
+  std::set<int> isolated_via_nodes;
+  omp_set_num_threads(ord::OpenRoad::openRoad()->getThreadCount());
+#pragma omp parallel for schedule(dynamic)
+  for (int i = 0; i < via_positions.size(); i++) {
+    bool already_visited = false;
+#pragma omp critical
+    {
+      if (visited[i]) {
+        already_visited = true;
+      }
+      visited[i] = true;
+    }
+    if (already_visited) {
+      continue;
+    }
+    std::vector<int> neighbors = tree.radiusSearch(via_positions[i], max_spc);
+
+// Check if there are neighbors other than the point itself
+#pragma omp critical
+    {
+      bool is_isolated = true;
+      for (const auto& neighbor : neighbors) {
+        if (neighbor != i) {
+          visited[neighbor] = true;
+          is_isolated = false;
+        }
+      }
+      if (is_isolated) {
+        isolated_via_nodes.insert(i);
+      }
+    }
```

Comment:
With visited being atomic you only need critical to cover `isolated_via_nodes.insert(i);` not the whole block.

Reply:
ok